pageserver/src/walredo.rs

```diff
@@ -226,71 +233,68 @@ impl PostgresRedoManager {
-        if process_guard.is_none() {
-            let p = PostgresRedoProcess::launch(self.conf, self.tenant_id, pg_version)?;
-            *process_guard = Some(p);
+            //
+            // Launch the WAL redo process on first use
+            //
+            // TODO-1: it requires locking mutex oin each redo request. It will be better to launch process at
+            // PostgresRedoManage::new but then we should find a way to prevent laucnhing walredo process on
+            // bootsrap incation of pageserver, otherwise two initdb may confict with ah other.
+            //
+            // TODO-2: we are not reading stderr of walredo process right now and so not handling waredo errors.
+            //
+            if process_guard.is_none() {
+                let p = PostgresRedoProcess::launch(self.conf, self.tenant_id, pg_version)?;
+                *process_guard = Some(p);
+            }
         }
-        let process = process_guard.as_mut().unwrap();
-
-        WAL_REDO_WAIT_TIME.observe(lock_time.duration_since(start_time).as_secs_f64());
-
-        // Relational WAL records are applied using wal-redo-postgres
-        let buf_tag = BufferTag { rel, blknum };
-        let result = process
-            .apply_wal_records(buf_tag, base_img, records, wal_redo_timeout)
-            .map_err(WalRedoError::IoError);
-
-        let end_time = Instant::now();
-        let duration = end_time.duration_since(lock_time);
-
-        let len = records.len();
-        let nbytes = records.iter().fold(0, |acumulator, record| {
-            acumulator
-                + match &record.1 {
-                    NeonWalRecord::Postgres { rec, .. } => rec.len(),
-                    _ => unreachable!("Only PostgreSQL records are accepted in this batch"),
-                }
-        });
-
-        WAL_REDO_TIME.observe(duration.as_secs_f64());
-        WAL_REDO_RECORDS_HISTOGRAM.observe(len as f64);
-        WAL_REDO_BYTES_HISTOGRAM.observe(nbytes as f64);
-
-        debug!(
-            "postgres applied {} WAL records ({} bytes) in {} us to reconstruct page image at LSN {}",
-            len,
-            nbytes,
-            duration.as_micros(),
-            lsn
-        );
+        let tag = BufferTag { rel, blknum };
+        let mut writebuf: Vec<u8> = Vec::with_capacity((BLCKSZ as usize) * 3);
+        build_begin_redo_for_block_msg(tag, &mut writebuf);
+        if let Some(img) = base_img {
+            build_push_page_msg(tag, &img, &mut writebuf);
+        }
+        for (lsn, rec) in records.iter() {
+            if let NeonWalRecord::Postgres {
+                will_init: _,
+                rec: postgres_rec,
+            } = rec
+            {
+                build_apply_record_msg(*lsn, postgres_rec, &mut writebuf);
+            } else {
+                return Err(WalRedoError::IoError(Error::new(
+                    ErrorKind::Other,
+                    "tried to pass neon wal record to postgres WAL redo",
+                )));
```

Comment:
An idea, maybe later: Might be a good idea to test if there's any need to parse and apply wal records in rust or just apply all of them in postgres.

Or do we need some currently implemented custom handling for some records?

Reply:
Very few records are not handled  by pageserver itself. So it seems to be reasonable to assume that any page will require some WAL records which need to be processed by wal redo process. So I do not see much sense in optimizing this "if".
Or may be I do not understand your proposal...

Reply:
The idea was to not process any wal records in rust instead to offload all of them to the walredo process, since this IPC is now becoming free.

Reply:
We are processing some WAL records at Neon side not only because it is more efficient than sending them to walredo process, but just because we need to do more than Postgres does. For example we need to track transaction commits and maintain CLOG. Postgres certainly also updates CLOG during recovery. But SLRUs are not currently maintained by buffer manager and so has to be accessed in different way.

Reply:
Forgot respond anything but understood now, also Stas explained a bit about why so resolving.